File: test files/typescript/senior_iniciante.ts
type TaskStatus = 'todo' | 'doing' | 'done' | 'blocked';
type TaskTag = 'critical' | 'customer-facing';

type CompleteTaskFailure =
  | 'TASK_NOT_FOUND'
  | 'TASK_WITHOUT_OWNER'
  | 'TASK_BLOCKED'
  | 'INVALID_ESTIMATE'
  | 'VERSION_CONFLICT'
  | 'AUDIT_UNAVAILABLE';

type Result<T, E extends string> =
  | { ok: true; data: T }
  | { ok: false; error: E };

type TaskId = number & { readonly __brand: 'TaskId' };
type ActorId = number & { readonly __brand: 'ActorId' };
type TaskVersion = number & { readonly __brand: 'TaskVersion' };

type CompletionCommand = Readonly<{
  taskId: TaskId;
  expectedVersion: TaskVersion;
  actorId: ActorId;
}>;

type TaskSnapshot = Readonly<{
  id: TaskId;
  title: string;
  ownerId: number | null;
  status: TaskStatus;
  estimateHours: number;
  version: TaskVersion;
  tags: ReadonlyArray<TaskTag>;
}>;

interface TaskRepository {
  findById(id: TaskId): Promise<TaskSnapshot | null>;
  save(task: TaskSnapshot): Promise<TaskSnapshot>;
}

interface AuditGateway {
  publish(entry: AuditEntry): Promise<void>;
}

interface DomainEventBus {
  publish(event: DomainEvent): Promise<void>;
}

interface MonitoringGateway {
  warn(event: string, context: Record<string, unknown>): void;
  info(event: string, context: Record<string, unknown>): void;
}

interface Clock {
  now(): Date;
}

type AuditEntry = {
  entityId: TaskId;
  action: 'task.completed';
  at: string;
  metadata: {
    actorId: ActorId;
    previousStatus: TaskStatus;
    previousVersion: number;
    estimateHours: number;
    tags: ReadonlyArray<TaskTag>;
  };
};

type DomainEvent = Readonly<{
  type: 'task.completed';
  entityId: TaskId;
  occurredAt: string;
  actorId: ActorId;
  payload: Readonly<{
    previousVersion: number;
    nextVersion: number;
  }>;
}>;

type CompletionArtifacts = Readonly<{
  auditEntry: AuditEntry;
  domainEvent: DomainEvent;
}>;

type MonitoringContext = Readonly<{
  taskId: TaskId;
  actorId: ActorId;
  tags: ReadonlyArray<TaskTag>;
  previousVersion: TaskVersion;
  nextVersion?: TaskVersion;
  expectedVersion?: TaskVersion;
  currentVersion?: TaskVersion;
}>;

class TaskCompletionPolicy {
  evaluate(
    task: TaskSnapshot
  ): Result<TaskSnapshot, Extract<CompleteTaskFailure, 'TASK_WITHOUT_OWNER' | 'TASK_BLOCKED' | 'INVALID_ESTIMATE'>> {
    if (task.ownerId === null) {
      return { ok: false, error: 'TASK_WITHOUT_OWNER' };
    }

    if (task.status === 'blocked') {
      return { ok: false, error: 'TASK_BLOCKED' };
    }

    if (task.estimateHours <= 0) {
      return { ok: false, error: 'INVALID_ESTIMATE' };
    }

    return { ok: true, data: task };
  }
}

class CompleteTaskUseCase {
  constructor(
    private readonly taskRepository: TaskRepository,
    private readonly auditGateway: AuditGateway,
    private readonly domainEventBus: DomainEventBus,
    private readonly monitoringGateway: MonitoringGateway,
    private readonly clock: Clock,
    private readonly policy: TaskCompletionPolicy
  ) {}

  async execute(
    command: CompletionCommand
  ): Promise<Result<{ taskId: TaskId; completedAt: string; nextVersion: number }, CompleteTaskFailure>> {
    const task = await this.taskRepository.findById(command.taskId);
    if (!task) {
      this.monitoringGateway.warn('task.complete.not_found', { taskId: command.taskId, actorId: command.actorId });
      return { ok: false, error: 'TASK_NOT_FOUND' };
    }

    const validation = this.policy.evaluate(task);
    if (!validation.ok) {
      this.monitoringGateway.warn('task.complete.validation_failed', {
        ...this.createMonitoringContext(task, command.actorId),
        code: validation.error,
      });
      return validation;
    }

    if (task.version !== command.expectedVersion) {
      this.monitoringGateway.warn('task.complete.version_conflict', {
        ...this.createMonitoringContext(task, command.actorId),
        expectedVersion: command.expectedVersion,
        currentVersion: task.version,
      });
      return { ok: false, error: 'VERSION_CONFLICT' };
    }

    const completedAt = this.clock.now().toISOString();
    const updatedTask = this.buildUpdatedTask(task);
    const completionArtifacts = this.buildCompletionArtifacts({
      command,
      task,
      updatedTask,
      completedAt,
    });

    try {
      await this.taskRepository.save(updatedTask);
      await this.publishCompletionArtifacts(completionArtifacts);
    } catch {
      this.monitoringGateway.warn('task.complete.audit_unavailable', {
        ...this.createMonitoringContext(task, command.actorId, updatedTask.version),
      });
      return { ok: false, error: 'AUDIT_UNAVAILABLE' };
    }

    this.monitoringGateway.info('task.complete.succeeded', {
      ...this.createMonitoringContext(task, command.actorId, updatedTask.version),
    });

    return {
      ok: true,
      data: {
        taskId: updatedTask.id,
        completedAt,
        nextVersion: updatedTask.version,
      },
    };
  }

  private buildUpdatedTask(task: TaskSnapshot): TaskSnapshot {
    return {
      ...task,
      status: 'done',
      version: /** @type {TaskVersion} */ (task.version + 1),
    };
  }

  private buildCompletionArtifacts({
    command,
    task,
    updatedTask,
    completedAt,
  }: {
    command: CompletionCommand;
    task: TaskSnapshot;
    updatedTask: TaskSnapshot;
    completedAt: string;
  }): CompletionArtifacts {
    return {
      auditEntry: {
        entityId: updatedTask.id,
        action: 'task.completed',
        at: completedAt,
        metadata: {
          actorId: command.actorId,
          previousStatus: task.status,
          previousVersion: task.version,
          estimateHours: task.estimateHours,
          tags: task.tags,
        },
      },
      domainEvent: {
        type: 'task.completed',
        entityId: updatedTask.id,
        occurredAt: completedAt,
        actorId: command.actorId,
        payload: {
          previousVersion: task.version,
          nextVersion: updatedTask.version,
        },
      },
    };
  }

  private createMonitoringContext(
    task: TaskSnapshot,
    actorId: ActorId,
    nextVersion?: TaskVersion
  ): MonitoringContext {
    return {
      taskId: task.id,
      actorId,
      tags: task.tags,
      previousVersion: task.version,
      ...(nextVersion ? { nextVersion } : {}),
    };
  }

  private async publishCompletionArtifacts(artifacts: CompletionArtifacts): Promise<void> {
    await this.auditGateway.publish(artifacts.auditEntry);
    await this.domainEventBus.publish(artifacts.domainEvent);
  }
}

File: test files/java/senior_iniciante.java
import java.time.Clock;
import java.time.Instant;
import java.util.Map;
import java.util.Optional;

public class SeniorTaskCompletionService {

    private final TaskRepository taskRepository;
    private final TaskPolicy taskPolicy;
    private final AuditGateway auditGateway;
    private final MonitoringGateway monitoringGateway;
    private final Clock clock;

    public SeniorTaskCompletionService(
            TaskRepository taskRepository,
            TaskPolicy taskPolicy,
            AuditGateway auditGateway,
            MonitoringGateway monitoringGateway,
            Clock clock
    ) {
        this.taskRepository = taskRepository;
        this.taskPolicy = taskPolicy;
        this.auditGateway = auditGateway;
        this.monitoringGateway = monitoringGateway;
        this.clock = clock;
    }

    public CompletionResult complete(CompletionCommand command) {
        Optional<TaskSnapshot> maybeTask = taskRepository.findById(command.taskId());
        if (maybeTask.isEmpty()) {
            monitoringGateway.warn("task.complete.not_found", Map.of("taskId", command.taskId()));
            return CompletionResult.failure("TASK_NOT_FOUND");
        }

        TaskSnapshot task = maybeTask.get();
        PolicyDecision decision = taskPolicy.evaluate(task);
        if (!decision.allowed()) {
            monitoringGateway.warn(
                    "task.complete.validation_failed",
                    Map.of("taskId", task.id(), "code", decision.code())
            );
            return CompletionResult.failure(decision.code());
        }

        if (command.expectedVersion() != task.version()) {
            monitoringGateway.warn(
                    "task.complete.version_conflict",
                    Map.of("taskId", task.id(), "expectedVersion", command.expectedVersion(), "currentVersion", task.version())
            );
            return CompletionResult.failure("VERSION_CONFLICT");
        }

        Instant completedAt = clock.instant();
        TaskSnapshot updatedTask = task.completeAt(completedAt);
        taskRepository.save(updatedTask);

        auditGateway.publish(
                new AuditEntry(
                        updatedTask.id(),
                        "task.completed",
                        completedAt,
                        Map.of(
                                "previousStatus", task.status(),
                                "previousVersion", task.version(),
                                "estimateHours", task.estimateHours()
                        )
                )
        );

        monitoringGateway.info(
                "task.complete.succeeded",
                Map.of("taskId", updatedTask.id(), "nextVersion", updatedTask.version())
        );

        return CompletionResult.success(updatedTask.id(), completedAt, updatedTask.version());
    }

    public record CompletionCommand(long taskId, int expectedVersion) {
    }

    public record CompletionResult(boolean success, String code, Long taskId, Instant completedAt, Integer nextVersion) {
        public static CompletionResult success(long taskId, Instant completedAt, int nextVersion) {
            return new CompletionResult(true, "OK", taskId, completedAt, nextVersion);
        }

        public static CompletionResult failure(String code) {
            return new CompletionResult(false, code, null, null, null);
        }
    }

    public record TaskSnapshot(
            long id,
            String title,
            Long ownerId,
            String status,
            int estimateHours,
            int version
    ) {
        public TaskSnapshot completeAt(Instant completedAt) {
            return new TaskSnapshot(id, title, ownerId, "DONE", estimateHours, version + 1);
        }
    }

    public record PolicyDecision(boolean allowed, String code) {
        public static PolicyDecision allow() {
            return new PolicyDecision(true, "OK");
        }

        public static PolicyDecision deny(String code) {
            return new PolicyDecision(false, code);
        }
    }

    public interface TaskRepository {
        Optional<TaskSnapshot> findById(long taskId);

        void save(TaskSnapshot task);
    }

    public interface TaskPolicy {
        PolicyDecision evaluate(TaskSnapshot task);
    }

    public interface AuditGateway {
        void publish(AuditEntry entry);
    }

    public interface MonitoringGateway {
        void info(String event, Map<String, Object> context);

        void warn(String event, Map<String, Object> context);
    }

    public record AuditEntry(long entityId, String action, Instant occurredAt, Map<String, Object> metadata) {
    }
}
